clang-tools-extra/clang-tidy/daedalean/LambdaReturnTypeCheck.cpp

```diff
@@ -0,0 +1,42 @@
+//===--- LambdaReturnTypeCheck.cpp - clang-tidy ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "LambdaReturnTypeCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+void LambdaReturnTypeCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(lambdaExpr().bind("x"), this);
+}
+
+void LambdaReturnTypeCheck::check(const MatchFinder::MatchResult &Result)  {
+  const auto *MatchedDecl = Result.Nodes.getNodeAs<LambdaExpr>("x");
+  if (MatchedDecl->hasExplicitResultType())
+    return;
+
+  if (MatchedDecl->isGenericLambda())
+    return;
+
+  if (MatchedDecl->getCallOperator()->getReturnType()->isVoidType())
+    return;
```

Comment:
So lambdas that return void don't need to be explicitly declared as such? Should we change the rule so that it is required? (Why not?)

Reply:
better discuss it in coding standards cl

Reply:
OK. Let's land as is and adapt if we make this change in the coding standard.